hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/batch/TransactionHashBatchInserter.java

```diff
@@ -0,0 +1,234 @@
+import lombok.CustomLog;
+import lombok.Data;
+import lombok.SneakyThrows;
+import org.springframework.jdbc.datasource.DataSourceUtils;
+import org.springframework.transaction.support.TransactionSynchronization;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+@CustomLog
+public class TransactionHashBatchInserter implements BatchPersister {
+    private final Map<Integer, BatchInserter> shardBatchInserters = new ConcurrentHashMap<>();
+    private final DataSource dataSource;
+    private final MeterRegistry meterRegistry;
+    private final CommonParserProperties commonParserProperties;
+    private final Timer insertDurationMetric;
+    private final String shardedTableName;
+
+    private final Scheduler scheduler;
+    private final Map<String, ThreadState> threadConnections = new ConcurrentHashMap<>();
+
+    public TransactionHashBatchInserter(DataSource dataSource, MeterRegistry meterRegistry,
+                                        CommonParserProperties commonParserProperties) {
+        this.dataSource = dataSource;
+        this.meterRegistry = meterRegistry;
+        this.commonParserProperties = commonParserProperties;
+        this.shardedTableName = CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE,
+                TransactionHash.class.getSimpleName() + "_sharded");
+        this.insertDurationMetric = Timer.builder("hedera.mirror.importer.parse.insert")
+                .description("Time to insert transactions into sharded table").tag("table", this.shardedTableName)
+                .register(meterRegistry);
+        this.scheduler = Schedulers.newParallel(this.shardedTableName + "_shard_inserter", 8);
+    }
+
+    @Override
+    public void persist(Collection<?> items) {
+        threadConnections.clear();
+
+        if (items == null || items.isEmpty()) {
+            return;
+        }
+
+        try {
+            Stopwatch stopwatch = Stopwatch.createStarted();
+
+            Map<Integer, List<TransactionHash>> shardedItems = items.stream()
+                    .map(TransactionHash.class::cast)
+                    .collect(Collectors.groupingBy(TransactionHash::calculateV1Shard));
+
+            // After parser transaction completes, process all transactions for shards
+            addTransactionSynchronization(items);
+
+            Mono.when(shardedItems.entrySet()
+                            .stream()
+                            .map(shard -> this.processShard(shard).subscribeOn(scheduler))
+                            .toList())
+                    .block();
```

Comment:
Think you can simplify by combining shardItems and this to use grouped flux.

```suggestion
            // After parser transaction completes, process all transactions for shards
            addTransactionSynchronization(items);

            Flux.fromIterable(items)
                .groupBy(TransactionHash::calculateV1Shard)                
                .subscribeOn(Schedulers.parallel())
                .flatMap(Flux::collectList)
                .doOnNext(this::processShard)
                .subscribeOn(scheduler)
                .block();
```

Or can do collectList inside processShard.

Reply:
Doing it this way doesn't parallelize the work across the scheduler correctly. I did some playing around and I couldn't get a grouped flux to execute how we would want.

Reply:
We would have to process the grouped flux returned by groupBy by first calling a collect list on each of the fluxes so we can get the items for the shards. I attempted to do this and return a mono from that with subscribing on the scheduler but it doesn't seem to parallelize correctly.